### ISSUE TITLE
Forms monitoring: Datadog monitor to alarm if 500 errors ("Invalid PDF link") pass threshholds

### ISSUE THREAD

**Issue by jilladams** [department-of-veterans-affairs/va.gov-cms#15377]
## User Story or Problem Statement
AS A Public Websites maintainer
I WANT to be proactively notified when users are getting "invalid PDF" messages in the UI
SO THAT we can triage causes and make sure Veterans successfully download files / can access forms.

## Description or Additional Context 
https://dsva.slack.com/archives/C52CL1PKQ/p1695660801315459

Friday 9/22 from 1:30-2:30pm UTC, Forms had a spike in 500 "Bad PDF link" errors where form PDF files couldn't be downloaded, cause still TBD.
- Could be caused by vets-api latency, if that was happening -- [Platform support request](https://dsva.slack.com/archives/CBU0KDSB1/p1695660069145739) is open, to investigate.
- Not a Lighthouse problem (calls to their API were successful, [Thread with Lighthouse](https://dsva.slack.com/archives/CUB5X5MGF/p1695657274766899))
- Doesn't appear to be a Forms migration / filename changes issue. ([PW Monitoring channel thread](https://dsva.slack.com/archives/C05THHJHH2R/p1695649320070549))

**Sentry** monitors for these issues but cannot alarm us.
**Datadog** could alarm us, if we can configure a monitor to catch this case.

## Monitor criteria
**[Example event](http://sentry.vfs.va.gov/organizations/vsp/discover/platform-web:d60f6507ce2449a096d03ac56f36a558/?display=default&end=2023-09-22T14%3A59%3A59.000&field=title&field=event.type&field=project&field=user.display&field=timestamp&id=19&name=Find+Forms+errors&project=4&query=Find+Forms+event.type%3Adefault&sort=-timestamp&start=2023-09-22T12%3A00%3A00.000&widths=625&widths=-1&widths=-1&widths=-1&widths=-1)** that we would want to Monitor for:  500 errors on fetch HEAD [filepath].

Warn / Alert threshholds can match the ones we set in [Sentry alert](http://sentry.vfs.va.gov/organizations/vsp/alerts/metric-rules/platform-web/7/) that don't alert: 
* Warn: >20 in 30minutes
* Alert: > 50 in 30minutes
* Resolved: <10 in 30mintes 

## Acceptance Criteria
- [ ] A Datadog monitor has been created to alarm to #public-websites-monitoring if 500 errors happen over thresholds when fetching Forms PDF filepaths

**Comment by jilladams** [comment]
Noting: we need to understand if, when a filename changes and users get a 404 on old PDF links: what does that look like in Sentry / Datadog? Would those dashboards actually show a 404, or would it look like a 500 inability to download the file?

Also noting: filepaths will change in realtime but migrations run nightly so URLs in the UI will be udpated along with the migration.

**Comment by jilladams** [comment]
How does Sentry know about these errors? Wes reminded me:  FE JS is reporting these calls / responses to Sentry. So: Datadog would probably not be able to know about this, unless the FE sent the data to Datadog.

**Comment by chriskim2311** [comment]
This one is pretty tricky, I have figured out how sentry logs failed events through FE web components but for datadog it wont be as easy to monitor these issues. A synthetic test could catch latency with one failed test but it wont catch an accumulated number from real users. I am still investigating the best way forward to achieve that.

**Comment by chriskim2311** [comment]
Sentry catches these errors in the FE component flow.

Download URL HEAD response returns an error
`  'Find Forms - Form Detail - onDownloadLinkClick function error',`

In addition, we can catch these messages based on certain logic relating to a valid PDF and valid PDF URL if previous API call was valid. 

```
    let errorMessage = 'Find Forms - Form Detail - invalid PDF accessed';

    if (netWorkRequestError)
      errorMessage =
        'Find Forms - Form Detail - onDownloadLinkClick function error';

    if (!formPdfIsValid && !formPdfUrlIsValid)
      errorMessage =
        'Find Forms - Form Detail - invalid PDF accessed & invalid PDF link';

    if (formPdfIsValid && !formPdfUrlIsValid)
      errorMessage = 'Find Forms - Form Detail - invalid PDF link';
```

**Comment by jilladams** [comment]
Part of the reason we were hoping to get this working in Datadog is that Sentry alarms don't work for us. 
* Alert we created last year: [Forms errors warning / critical notifications](http://sentry.vfs.va.gov/organizations/vsp/alerts/metric-rules/platform-web/7/)
* Plat ticket to get Sentry > Slack working (has been in eval / progress since ~April 2020): https://github.com/department-of-veterans-affairs/va.gov-team/issues/55091#issuecomment-1483485705

**Comment by jilladams** [comment]
Mid-sprint: Blocked, Datadog has committed to prioritize this work.

**Comment by chriskim2311** [comment]
Datadog support ticket: help.ddog-gov.com/hc/requests/2891

**Comment by jilladams** [comment]
@chriskim2311 2 things: 
1. Saw your support ticket got closed - wondering if that actually resolved the Sentry piece? 
2. Poking around today and noticed that in Sentry there's a "Missing trace" not on the top right sidebar if you click through to any single failure, e.g. http://sentry.vfs.va.gov/organizations/vsp/discover/platform-web:361b350410ca45f297b9646c20b32975/?display=default&field=title&field=event.type&field=project&field=user.display&field=timestamp&id=19&name=Find+Forms+errors&project=4&query=Find+Forms+event.type%3Adefault&sort=-timestamp&statsPeriod=24h&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1
![Screenshot 2023-10-04 at 2 33 04 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/d2a62496-9da0-4c5d-ac04-fe4c908eb5fa)


The docs linked to there: https://docs.sentry.io/platforms/javascript/performance/

The nature of the failures we see in Sentry is pretty mysterious to me. We know we get onClick download errors, and we know we get invalid PDF links, but I'm not sure we have a really strong understanding of the conditions that cause those errors. (I could be wrong - lemme know if you do.) Wondering if getting set up to get the quick traces might help?  This might be a bunny trail that needs its own ticket. Lemme know what you think.

**Comment by chriskim2311** [comment]
@jilladams Thanks for the comments! I have been unblocked from datadog and obtained admin access for Sentry. But still need to have data flowing into datadog which I am working on.

We do know what causes each of the errors in the FE code the errors will occur in this order. 

1. `Fetch API for forms data` continue if valid or set `formPdfIsValid` to false.
2. `Fetch API for valid PDF` continue if valid or set `formPdfUrlIsValid` to false.


**Find Forms - Form Detail - onDownloadLinkClick function error** (Any API error from above)
- When a user clicks on the download PDF button and the API call to see if the download URL is valid is not successful

**Find Forms - Form Detail - invalid PDF accessed** (`formPdfIsValid = false and formPdfUrlIsValid = true `)
- When a user accesses an invalid PDF but the URL is valid

**Find Forms - Form Detail - invalid PDF link**  (`formPdfIsValid = true and formPdfUrlIsValid = false `)
- When a user accesses a valid PDF with an invalid URL

**Comment by jilladams** [comment]
Talked to Chris to try and better understand what this means, interms of: who's at fault where in the pipeline, and what the conditions mean in terms of how the APIs get called.   

* We call vets-api (v0/forms) to see if the form exists. `Fetch API for forms data`:
    * If we get an API response with specific metadata in it: `formPdfIsValid` = true
    * If we get an API response, but not the right metadata: `formPdfIsValid` = false 
    * If we get an API error response, we fail out, and report to Sentry: **Find Forms - Form Detail - onDownloadLinkClick function error** 
      * We don't always know _why_ this happens. 
      * Most seen error in Sentry 
* If we didn't error out (if `formPdfIsValid` = true OR false), we use the data we received in the first call, and call Lighthouse with it, to determine if there's a valid PDF at the URL we recieved from vets-api: `Fetch API for valid PDF` continue if valid or set `formPdfUrlIsValid` to false.
  * If `formPdfIsValid` = true, and Lighthouse returns a URL: `formPdfUrlIsValid`= true: user gets to download the form
  * If `formPdfIsValid` = true, LH returns an error: `formPdfUrlIsValid`= false: we return  **Find Forms - Form Detail - invalid PDF link** to Sentry
    * This can be when FormsDB server is down, or Forms Mgrs changed a PDF filename, and the CMS hasn't caught up with overnight data sync.
  * If `formPdfIsValid` = false, and `formPdfUrlIsValid` = true: we send **Find Forms - Form Detail - invalid PDF accessed** to Sentry
    - When a user accesses an invalid PDF but the URL is valid. This is weird and we can't find examples of this happening. 

 @chriskim2311 lemme know if I got this right?

**Comment by chriskim2311** [comment]
@jilladams Looks good on my end!

**Comment by chriskim2311** [comment]
Status for the end of sprint. I am currently waiting to hear back from platform and Sentry to see why there is a disconnect in the alert functionality on their end. I have been told it might be related to how alerts are set and how emails have not been going out when alert criteria has been met.

As of now I have set up the webhook integration for datadog in Sentry but events are not flowing to datadog for a new threshold monitor. This will most likely carry over to next sprint if not resolved by 10/10 tomorrow.  

Platform ticket Thread: https://dsva.slack.com/archives/CBU0KDSB1/p1696528379456509

**Comment by jilladams** [comment]
Just noting high level since I was thinking through whether this is all useful work given the blockers:
* Sentry tracks failures currently, we can manually monitor that, but don't get alerted. And: the dashboard we do have is a slushpile of a variety of errors, so it's hard to detect trends.
* The goal here was to 
  * get alerted by _something_ when error rates spike
  * See some more granular assessment of the various kinds of failures we've spelled out in the comments above, which Datadog seems to handle better.
*  if Sentry alerts get unstuck, that is a win even without Datadog integration. 
* Feeding Sentry into datadog seems like it will help with better dashboards / understanding of the why. (@chriskim2311 does that seem true? Or: mind noting what we will gain from the Sentry > Datadog integration?)

**Comment by chriskim2311** [comment]
@jilladams Correct! Sentry Alerts are being looked into as nothing is currently being sent for errors or issues on the FE. If that becomes unblocked I also see it as a win. Plat is also still looking into the Slack integration as well, but that will take some time. 

Feeding Sentry data to datadog will provide better dashboards and slack alerts are currently working so that can be used for error threshold monitor alerts.

**Comment by chriskim2311** [comment]
@jilladams For this ticket I have created monitors for all three sentry FE alerts. 

I also created a public websites datadog dashboard for easier access: https://app.ddog-gov.com/sb/f327ad72-c02a-11ec-a50a-da7ad0900007-442b3d138d37e02f299ac7a7847d7b84

Metrics for alerts are as follows for all three. These thresholds live in Sentry but they will trigger alerts in datadog to send to Slack. 

20 > 15 mins Warning
50 > 15mins Alert

With the new rate limit increases I believe we can account for that by raising the requirements to trigger these monitors. Let me know what you think!

**Comment by jilladams** [comment]
Gorgeous. Linking to the 3 Sentry-related monitors just for my own sake: 
[Sentry Forms: Find Forms - Form Detail - onDownloadLink Click Error Monitor
](https://vagov.ddog-gov.com/monitors/165098)
[Sentry Forms: Find Forms - Form Detail - invalid PDF link Error Monitor
](https://vagov.ddog-gov.com/monitors/167982)
[Sentry Forms: Find Forms - Form Detail - invalid PDF accessed Error Monitor
](https://vagov.ddog-gov.com/monitors/167983)

I'm content to close this out. Doing so! Great work, Chris.
